test/e2e/utils_test.go

```diff
@@ -167,3 +191,163 @@ func verifyDeploymentArgs(k8sclient *kubernetes.Clientset, deploymentName string
+// - certificate hasn't expired
+// - certificate has subject CN matching provided hostname
+func verifyCertificate(ctx context.Context, secretName, namespace, hostname string) (bool, error) {
+	var isVerified bool
+	err := wait.PollImmediate(PollInterval, TestTimeout, func() (bool, error) {
+		secret, err := loader.KubeClient.CoreV1().Secrets(namespace).Get(ctx, secretName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		isVerified, err = library.VerifyCertificate(secret, hostname)
+		if err != nil {
+			return true, err
+		}
+		return true, nil
+	})
+	return isVerified, err
+}
+
+// verifyCertificateRenewed repeatedly loads the tls secret as a X509 certificate every pollDuration
+// and returns true if certificate was renewed at least once
+func verifyCertificateRenewed(ctx context.Context, secretName, namespace string, pollDuration time.Duration) (bool, error) {
+	var wasRenewedAtleastOnce bool
+	var initExpiryTime *time.Time
+	err := wait.PollImmediate(pollDuration, TestTimeout, func() (bool, error) {
+		secret, err := loader.KubeClient.CoreV1().Secrets(namespace).Get(ctx, secretName, metav1.GetOptions{})
+		if err != nil {
+			return false, nil
+		}
+
+		cert, err := library.GetX509Certificate(secret)
+		if err != nil {
+			return false, err
+		}
+
+		// cert expiry time is recorded upon initial run
+		if initExpiryTime == nil {
+			initExpiryTime = &cert.NotAfter
+		}
+
+		// checks if expiry time was updated
+		if *initExpiryTime == cert.NotAfter {
+			return false, nil
+		}
+
+		// iff expiry time was updated, check if new expiry is not ahead
+		// return an error, else certificate was renewed properly
+		if !cert.NotAfter.After(*initExpiryTime) {
+			return false, fmt.Errorf("previous expiry time of the certificate cannot be ahead of the current expiry time")
+		}
+
+		wasRenewedAtleastOnce = true
+		return true, nil
+	})
+	return wasRenewedAtleastOnce, err
+}
+
+func resetCertManager(ctx context.Context, client *certmanoperatorclient.Clientset, loader library.DynamicResourceLoader) error {
+	// update operator.openshift.io/v1alpha1 CertManager object
+	// with default values
+	_, err := client.OperatorV1alpha1().CertManagers().Update(ctx, &defaultCertManagerState, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// remove any entries present in Subscription spec.config for
+	// user provided injected env vars, etc.
+	// it should put operator back to default deployment
+	subName, err := getCertManagerOperatorSubscription(ctx, loader)
+	if err != nil {
+		return err
+	}
+
+	// to get an empty spec.config
+	patch := map[string]interface{}{
+		"spec": map[string]interface{}{
+			"config": nil,
+		},
+	}
+	payload, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+
+	subscriptionClient := loader.DynamicClient.Resource(subscriptionSchema).Namespace("cert-manager-operator")
+	_, err = subscriptionClient.Patch(ctx, subName, types.MergePatchType, payload, v1.PatchOptions{})
+	return err
```

Comment:
Better to add this code within retryOnConflict.

Reply:
I ended up removing the .Update() call to certmanager.operator object, because it was incorrectly used here (without a prior .Get, erroring out the update, failing the e2e ultimately).

Instead, it felt more clean to use a .Patch() call with the certmanager object (so we don't have to Get, set the specific fields to nil, and then perform... i.e. more code to maintain for future API changes). @arkadeepsen do you still think I should use `retry.retryOnConflict` with .Patch() calls also (IMHO, I saw in most places it's just paired with Update generally so unsure if Patch also needs it).

@thejasn it would be seemingly like (this is in favour of least maintainability of `resetCertManager` state):
```golang
// delete the operator.openshift.io/v1alpha1 CertManager object,
// it should get created again with default values,
// remove the finalizers before deleting the object
finalizerPatch := []interface{}{
	map[string]interface{}{
		"op":   "remove",
		"path": "/metadata/finalizers",
	},
}
payload, err := json.Marshal(finalizerPatch)
if err != nil {
	return err
}

_, err = client.OperatorV1alpha1().CertManagers().Patch(ctx, "cluster", types.JSONPatchType, payload, metav1.PatchOptions{})
if err != nil {
	return err
}
``` 
T'is the one that works,
- .Update() needs me to set each API field within spec to null/{} manually
- simple .Delete() get stalls after a few repeated calls due to unclean finalizer cleanup sometimes
-  this one works well across multiple repeated runs (I shall wait for e2e to pass and verify the same)

Let me know :)

Reply:
With `Patch` `retry.retryOnConflict` is probably not required.

Reply:
The patch + delete isn't working well, the operator status on the certmanager object doesn't reconcile back properly, ends up failing the test.
So I'll switch back to Get + Update with a retry.retryOnConflict wrapped around.